### PR TITLE
Supports boolean and number labels. Fixes #1598

### LINKED ISSUE #1598: Add support for numeric and boolean labels
**Is your feature request related to a problem? Please describe.**

Since APM Server 6.7, labels can have numerical or boolean values. We should add support for these in the agent's API, which currently supports only string values (non-string values are stringified).

**Describe the solution you'd like**

[`setLabel`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/transaction-api.html#transaction-set-label) and [`addLabels`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/transaction-api.html#transaction-add-labels) should accept numeric and boolean values, and should not stringify them.

Just changing the behaviour would be a breaking change (particularly because they would end up as different Elasticsearch field types, causing indexing to fail), so that change would have to be done with a major version bump.

**Describe alternatives you've considered**

If waiting for a major version bump is too much, then we could add parameters to the functions that instruct them to not stringify values. Those parameters would be removed in a future major version, and the behaviour would be switched so stringification is not applied by default.

**Additional context**
https://www.elastic.co/guide/en/apm/get-started/current/metadata.html#labels-fields

### PR DESCRIPTION
**Pull request by ddolcimascolo**
Support boolean and numeric labels.

Fixes #1598 

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update documentation
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/master/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Comment by trentm** [comment]
Thanks! I'll take a look soon.

**Comment by trentm** [comment]
jenkins run the tests

File: docs/configuration.asciidoc [hunk 1]
```diff
@@ -834,6 +834,16 @@ Example:
...
+[[extended-label-types]]
+==== `extendedLabelTypes`
+
+* *Type:* Boolean
+* *Default:* `false`
+* *Env:* `ELASTIC_APM_EXTENDED_LABEL_TYPES`
+
+Whether numeric and boolean labels should be sent as-is to the APM server.
+When this option is `false`, all labels are sent as strings, no matter their original type.
+
```
Comment (trentm) [inline docs/configuration.asciidoc:846]: Did you consider adding an optional argument to `setLabel` and `addLabels`, something like:

```diff
-GenericSpan.prototype.setLabel = function (key, value) {
+// TODO: in v4, remove the stringify option and never stringify
+GenericSpan.prototype.setLabel = function (key, value, stringify = true) {
```

Pros that I see:
- The user of `setLabel()` needs to explicitly work through each usage in their code to verify that changing 'stringify' won't break them.
- One isn't led to infer that `extendedLabelTypes=false` might affect the `globalLabels` config var (it won't, globalLabels already passes through bools and numbers untouched).

Cons:
- The user may need to slog through multiple usages. :)

Comment (ddolcimascolo) [reply docs/configuration.asciidoc:846]: I didn't TBH. I very much like the way you can configure the agent with env vars, like most containing apps do anyway... I thought users could
1. Update the app environment, restart the app
2. Let the agent send a few transactions
3. Revert the environment, restart the app

Then they can safely test their label-specific ES queries on the new transactions and validate that the type changes work.

But I'm OK to move to an explicit third parameter and drop the config flag, you tell me 👍

Comment (trentm) [reply docs/configuration.asciidoc:846]: > But I'm OK to move to an explicit third parameter and drop the config flag

Yes please, and thanks. My initial gut reaction is to prefer the extra param to `setLabel` and `addLabels`. I asked a colleague and he preferred the extra param as well (more strongly than I).

File: CHANGELOG.asciidoc [hunk 2]
```diff
@@ -29,6 +29,20 @@ Notes:
...
+
+[float]
+===== Features
+
+* feat: support numeric and boolean labels {pull}1909[#1909] +
+  When `extendedLabelTypes` config option is set to `true`, boolean and numeric
+  labels are sent as-is to the APM server, wihtout stringification.
+
+[float]
+===== Bug fixes
```
[the remark this answers is not in the document]
Comment (ddolcimascolo) [reply CHANGELOG.asciidoc]: Done

**Comment by trentm** [comment]
jenkins run the tests

**Review by trentm — CHANGES_REQUESTED** [review]
Sorry again for the slow response (and ahead of time for after this because I'll be on break for the holidays for a bit).

1. I've made a couple suggestions below for a comment and the changelog.
2. I've updated the docs for this change. Can you please apply this patch to your branch: https://gist.github.com/trentm/1f1afdaa42661a05a6f9141864d2f718
3. Can you merge in changes from master?

**Comment by ddolcimascolo** [comment]
Hi @trentm 

I've applied your suggestions, rebased on master and merged the doc patch. Sorry I missed this :astonished: 

Regards,
David

**Comment by trentm** [comment]
jenkins run the tests

**Comment by trentm** [comment]
@elasticmachine, run elasticsearch-ci/docs

**Review by trentm — APPROVED** [review]
Because I just *did* a 3.10 release the changelog and a couple refs to 3.10 are out of date. However, I'm just going to merge and fix those up in a follow up commit to avoid the "dropping the ball"-potential of another back and forth.

Thanks!

**Comment by trentm** [comment]
There will be a minor release of the agent in the next two weeks that will include this.

**Comment by ddolcimascolo** [comment]
Thx for merging!

**Pull request merged**